services/port_manager/src/main/java/com/futurewei/alcor/portmanager/service/PortServiceImpl.java

```diff
@@ -208,25 +209,51 @@ public PortWebJson getPort(String projectId, String portId) throws Exception {
         return result;
     }
 
-    private List<NeighborEntry> buildNeighborTable(List<NeighborInfo> localNeighborInfos, List<NeighborInfo> l3Neighbors) {
-        List<NeighborEntry> neighborTable = new ArrayList<>();
+    private Map<String, List<NeighborEntry>> buildNeighborTable(List<NeighborInfo> localNeighborInfos, List<NeighborInfo> l3Neighbors) {
+        Map<String, List<NeighborEntry>> neighborTable = new HashMap<>();
         for (NeighborInfo localNeighborInfo: localNeighborInfos) {
+            List<NeighborEntry> neighborEntries = new ArrayList<>();
             for (NeighborInfo l3Neighbor: l3Neighbors) {
                 NeighborEntry neighborEntry = new NeighborEntry();
                 neighborEntry.setNeighborType(NeighborEntry.NeighborType.L3);
                 neighborEntry.setLocalIp(localNeighborInfo.getPortIp());
                 neighborEntry.setNeighborIp(l3Neighbor.getPortIp());
-                neighborTable.add(neighborEntry);
+                neighborEntries.add(neighborEntry);
             }
+
+            String portIp = localNeighborInfo.getPortIp();
+            neighborTable.put(portIp, neighborEntries);
         }
 
         return neighborTable;
     }
 
-    private List<NeighborEntry> updateNeighborTable(RouterUpdateInfo routerUpdateInfo, List<NeighborInfo> neighborInfos) throws CacheException {
+    private Map<String, List<NeighborEntry>> updateNeighborTable(RouterUpdateInfo routerUpdateInfo, Map<String, NeighborInfo> neighborInfos) throws CacheException {
         String vpcId = routerUpdateInfo.getVpcId();
         String subnetId = routerUpdateInfo.getSubnetId();
-        List<String> oldSubnetIds = routerUpdateInfo.getOldSubnetIds();
+        List<String> gatewayPortIds = routerUpdateInfo.getGatewayPortIds();
+        List<String> gatewaySubnetIds = new ArrayList<>();
+
+        Map<String, Object[]> filterParams = new HashMap<>();
+        filterParams.put("id", gatewayPortIds.toArray());
+        Map<String, PortEntity> gatewayPortEntities = portRepository.findAllPortEntities(filterParams);
+        if (gatewayPortEntities != null) {
+            for (Map.Entry<String, PortEntity> entry: gatewayPortEntities.entrySet()) {
+                PortEntity portEntity = entry.getValue();
+                List<PortEntity.FixedIp> fixedIps = portEntity.getFixedIps();
+                if (fixedIps == null) {
+                    LOG.warn("Can not find fixedIp of gateway port: {}", portEntity);
+                    continue;
+                }
+
+                for (PortEntity.FixedIp fixedIp: fixedIps) {
+                    String subnetId1 = fixedIp.getSubnetId();
+                    if (!StringUtils.isEmpty(subnetId1)) {
+                        gatewaySubnetIds.add(subnetId1);
+                    }
+                }
+            }
+        }
```

Comment:
what if gatewayPortEntities ==null here? Should we add some special handling?

Reply:
PortController will check its legitimacy.